### PR TITLE
Create a 'StrictProvenanceBackend' that restores CHERI functionality

### PR DESCRIPTION
**Pull request by rmn30**
This current draft duplicates way too much code and will have to be
revised.  Essential all it does is provide a custom SlabMetadata class
that extends FrontendSlabMetadata with a chunk_ptr to store the
original pointer to the slab with the original bounds of the mmaped
region (this used to exist but was removed during
refactoring). alloc_range sets this in the SlabMetadata and
dealloc_range uses it instead of the pointer argument (which might
have restricted bounds).

### CODE REVIEW

File: src/snmalloc/backend/strictprovenancebackend.h [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+        address_cast(alloc),
+        Pagemap::get_metaentry(address_cast(alloc)).get_slab_metadata());
+      Pagemap::set_metaentry(address_cast(alloc), size, t);
+
+      local_state.get_meta_range().dealloc_range(
+        capptr::Chunk<void>(&slab_metadata), sizeof(SlabMetadata));
+
+      // On CHERI platforms, we need to re-derive to get a pointer to
+      // the chunk.
+      auto diff = pointer_diff(slab_metadata.chunk_ptr, alloc);
```
Comment (mjp41) [inline src/snmalloc/backend/strictprovenancebackend.h:224]: `slab_metadata` deallocated by this point (UAF)

**Comment by nwf** [comment]
FWIW, the Morello cluster is very nearly ready to serve as a CI backend going forward; if you have an updated version of this you want me to try putting through that machinery, let me know.

File: src/snmalloc/backend/strictprovenancebackend.h [hunk 2]
```diff
@@ -0,0 +1,241 @@
+#if defined(SNMALLOC_CHECK_CLIENT) && !defined(OPEN_ENCLAVE)
+/**
```
Comment (davidchisnall) [inline src/snmalloc/backend/strictprovenancebackend.h:2]: We are missing a `#pragma once` at the start of this file.

File: src/snmalloc/backend/globalconfig.h [hunk 3]
```diff
@@ -21,18 +22,23 @@ namespace snmalloc
   }
 #  endif
 
+  using BackendAlloc = std::conditional_t<
```
Comment (davidchisnall) [inline src/snmalloc/backend/globalconfig.h:25]: Please add a doc comment for this.

File: src/snmalloc/backend/strictprovenancebackend.h [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+  public:
+    using Pagemap = BasicPagemap<
+      StrictProvenanceBackend,
+      PAL,
+      ConcretePagemap,
+      PageMapEntry,
+      false>;
+
+    static constexpr bool CONSOLIDATE_PAL_ALLOCS = false;
```
Comment (davidchisnall) [inline src/snmalloc/backend/strictprovenancebackend.h:54]: There are lots of missing doc comments here, but this one in particular is confusing.  Is it actually meant to be public or is it used only in the definition of the ranges below?

**Comment by mjp41** [comment]
Should this be closed now?

**Comment by rmn30** [comment]
I think it contains some changes to clear pointers that didn't make it into @nwf 's version so I should rebase on top of main and resubmit at some point.

**Comment by mjp41** [comment]
It looks like the CLEAN_POINTER changes all made it into #528, and the rest has been superseeded by #537, so I am going to close this.

**Pull request closed without merging**